core/controllers/base.py

```diff
@@ -341,9 +340,20 @@ def vaidate_and_normalize_args(self):
         payload_arg_keys = []
         request_arg_keys = []
         for arg in self.request.arguments():
-            if arg in self.ARGS_WHICH_DO_NOT_NEED_SCHEMA_VALIDATION:
+            if arg == 'csrf_token':
+                # 'csrf_token' has been already validated in the
+                # dispatch method.
                 continue
-            if arg == 'payload':
+            elif arg == 'source':
+                source_url = self.request.get('source')
+                regex_pattern = (
+                    r'http[s]?://(?:[a-zA-Z]|[0-9]|[$-_@.&+]|[!*\(\),]|(?:%[0-9a-fA-F][0-9a-fA-F]))+' # pylint: disable=line-too-long
```

Comment:
We know what all our source URLs look like and they aren't this complicated. Can we simplify?

Reply:
Done